### PR TITLE
Issue #3733: Modify indent check to handle multiple, stacked annotations

### PR DESCRIPTION
**Pull request by bamapookie**
Issue #3733

checkAnnotationIndentation has been modified to check one annotation at a time, and recurse to handle additional annotations.

### CODE REVIEW

**Comment by romani** [comment]
UTs are passed is good but not enough for such complicated Check.
Please provide regression diff testing. Especially focus to guava project.

Diff tool - https://github.com/checkstyle/contribution/tree/master/patch-diff-report-tool 
Please follow readme file , please do not hesitate to ask us any questions.

**Comment by bamapookie** [comment]
@romani I'm about to run the regression test.  I am leaving out all the ones with additional excludes, since they are already included.  Is that OK?

**Comment by bamapookie** [comment]
@romani Also, I am not clear what I am supposed to edit in my_check.xml.  I am leaving as is.

**Comment by bamapookie** [comment]
The checkstyle tester fails with the following error:

```
Running Checkstyle on src/main/java ... with excludes ,**/resources-noncompilable/**/*,,,**/test/tools/pack200/typeannos/TypeUseTarget.java,**/test/tools/pack200/typeannos/TypeUseTarget.java,**/test/java/awt/Focus/ModalDialogActivationTest/ModalDialogActivationTest.java,**/test/javax/xml/bind/jxc/8073519/**,**/test/sun/tools/jhsdb/**,,,**/pmd/pmd-java/src/test/**/*,**/lombok-ast/test/**/*,,**/hibernate-orm/documentation/**/*,, ,,,,**apache-ant/src/tests/**/*,apache-ant/src/etc/testcases/,,,**/resources/**/*,,,,,
[INFO] Error stacktraces are turned on.
[INFO] Scanning for projects...
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building sample 0.0.1-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 0.082 s
[INFO] Finished at: 2017-01-20T12:28:49-05:00
[INFO] Final Memory: 7M/309M
[INFO] ------------------------------------------------------------------------
[ERROR] Unknown lifecycle phase ",,,,**apache-ant/src/tests/**/*,apache-ant/src/etc/testcases/,,,**/resources/**/*,,,,,". You must specify a valid lifecycle phase or a goal in the format <plugin-prefix>:<goal> or <plugin-group-id>:<plugin-artifact-id>[:<plugin-version>]:<goal>. Available lifecycle phases are: validate, initialize, generate-sources, process-sources, generate-resources, process-resources, compile, process-classes, generate-test-sources, process-test-sources, generate-test-resources, process-test-resources, test-compile, process-test-classes, test, prepare-package, package, pre-integration-test, integration-test, post-integration-test, verify, install, deploy, pre-clean, clean, post-clean, pre-site, site, post-site, site-deploy. -> [Help 1]
org.apache.maven.lifecycle.LifecyclePhaseNotFoundException: Unknown lifecycle phase ",,,,**apache-ant/src/tests/**/*,apache-ant/src/etc/testcases/,,,**/resources/**/*,,,,,". You must specify a valid lifecycle phase or a goal in the format <plugin-prefix>:<goal> or <plugin-group-id>:<plugin-artifact-id>[:<plugin-version>]:<goal>. Available lifecycle phases are: validate, initialize, generate-sources, process-sources, generate-resources, process-resources, compile, process-classes, generate-test-sources, process-test-sources, generate-test-resources, process-test-resources, test-compile, process-test-classes, test, prepare-package, package, pre-integration-test, integration-test, post-integration-test, verify, install, deploy, pre-clean, clean, post-clean, pre-site, site, post-site, site-deploy.
	at org.apache.maven.lifecycle.internal.DefaultLifecycleExecutionPlanCalculator.calculateLifecycleMappings(DefaultLifecycleExecutionPlanCalculator.java:246)
	at org.apache.maven.lifecycle.internal.DefaultLifecycleExecutionPlanCalculator.calculateMojoExecutions(DefaultLifecycleExecutionPlanCalculator.java:217)
	at org.apache.maven.lifecycle.internal.DefaultLifecycleExecutionPlanCalculator.calculateExecutionPlan(DefaultLifecycleExecutionPlanCalculator.java:127)
	at org.apache.maven.lifecycle.internal.DefaultLifecycleExecutionPlanCalculator.calculateExecutionPlan(DefaultLifecycleExecutionPlanCalculator.java:145)
	at org.apache.maven.lifecycle.internal.builder.BuilderCommon.resolveBuildPlan(BuilderCommon.java:96)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:109)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
```

**Comment by bamapookie** [comment]
Fixed the error.  There is a small bug in checkstyle-tester.  See here: https://github.com/checkstyle/contribution/issues/149

**Comment by bamapookie** [comment]
The regression test shows no difference. 83611 files checked.  Each had 5875 unique rows and 5875 severity warnings.

**Comment by rnveach** [comment]
> The regression test shows no difference.

We ask that you post the report somewhere so we can also see it. You can always use github's gh_pages.

> I am leaving out all the ones with additional excludes, since they are already included.

I'm not sure I understand this. For Indentation, it is best to run on all projects as you never know where you'll find a difference in. Excludes for projects are usually for Java grammar we can't parse for one reason or another.

> I am not clear what I am supposed to edit in my_check.xml. I am leaving as is.

The check you want to check should be there.
In this case, remove `AbstractClassName` and add `Indentation`.

Properties for the check is a different matter. Each project has its own custom settings and sometimes violations messages are hard to understand when the properties are not set correctly for the project.
I would just use the default properties for now, unless @romani requests something different.

**Comment by bamapookie** [comment]
Thanks, @rnveach, for that clarification.  I will rerun and post the results as you described.

The ones I excluded were duplicates of others already uncommented in the file.  For instance, I left out
```
#checkstyle-with-excludes|git|https://github.com/checkstyle/checkstyle.git|master|**/checkstyle-with-excludes/src/test/**/*,**/checkstyle-with-excludes/src/it/resources/**/*,**/resources-noncompilable/**/*
```
because the line above
```
checkstyle|git|https://github.com/checkstyle/checkstyle.git|master|**/resources-noncompilable/**/*
```
was already included, and covered more files.  The same was true for `sevntu-checkstyle-with-excludes`, `findbugs`, and `hibernate-orm-with-excludes`.

I just noticed that `guava-mvnstyle` actually checks master instead of v18, so I am including it as well.

**Comment by bamapookie** [comment]
Regression test complete.  https://bamapookie.github.io/checkstyle_issue_3733/

**Comment by romani** [comment]
False positive: 
https://bamapookie.github.io/checkstyle_issue_3733/#A10
```
  		@Override
          @SuppressWarnings({ "unchecked" })
 		public S deepCopyNotNull(S value) {
  			return (S) SerializationHelper.clone( value );
  		}
```

Violation was removed, but code is not correct. This case should appear in UTs (all problems from regression testing should be added to UTs)

**Comment by bamapookie** [comment]
https://bamapookie.github.io/checkstyle_issue_3733/my_check.xml
https://bamapookie.github.io/checkstyle_issue_3733/projects-to-test-on.properties

I don't think that line is a false positive.  Line 58 is positioned with spaces and the surrounding lines have tabs in them.  Maybe I should rerun the check with tabs = 4 spaces?  Not sure if there is a configuration option for that.

**Comment by romani** [comment]
> Maybe I should rerun the check with tabs = 4 spaces? Not sure if there is a configuration option for that.

thanks for confirmation.
It is my bad, I always forget about this nuance. All sources are different in tab size, so change to tabs size could trigger just another differences.

**Comment by romani** [comment]
@bamapookie , we did few fixes for our regression tool, please use groovy variant of it.

**Comment by bamapookie** [comment]
I tried running the updated checkstyle tester groovy script.  I used the same my_check.xml file and the updated projects-to-be-tested.properties file (uncommented as before).  I am now getting the following errors:

```
[INFO] Generating "Checkstyle" report    --- maven-checkstyle-plugin:2.17
/Users/shawn.kovalchick/Repositories/contribution/checkstyle-tester/src/main/java/hibernate-orm/documentation/src/main/docbook/devguide/en-US/chapters/events/extras/JpaCallbacksExample.java:22:81: expecting RPAREN, found ';'
/Users/shawn.kovalchick/Repositories/contribution/checkstyle-tester/src/main/java/hibernate-orm/documentation/src/main/docbook/devguide/en-US/chapters/multitenancy/extras/MultiTenantConnectionProviderImpl-single-cp.java:53:2: expecting RCURLY, found '...'
/Users/shawn.kovalchick/Repositories/contribution/checkstyle-tester/src/main/java/hibernate-orm/documentation/src/main/docbook/devguide/en-US/chapters/pc/extras/NaturalIdLoading.java:13:17: unexpected token: ;
/Users/shawn.kovalchick/Repositories/contribution/checkstyle-tester/src/main/java/hibernate-orm/documentation/src/main/docbook/devguide/en-US/chapters/pc/extras/NaturalIdLoading.java:19:62: expecting IDENT, found ')'
/Users/shawn.kovalchick/Repositories/contribution/checkstyle-tester/src/main/java/hibernate-orm/documentation/src/main/docbook/devguide/en-US/chapters/pc/extras/SimpleNaturalIdLoading.java:8:17: unexpected token: ;
/Users/shawn.kovalchick/Repositories/contribution/checkstyle-tester/src/main/java/hibernate-orm/documentation/src/main/docbook/devguide/en-US/chapters/pc/extras/SimpleNaturalIdLoading.java:14:62: expecting IDENT, found ')'
/Users/shawn.kovalchick/Repositories/contribution/checkstyle-tester/src/main/java/hibernate-orm/documentation/src/main/docbook/devguide/en-US/chapters/query_criteria/extras/select_wrapper.java:8:2: expecting RCURLY, found '...'
/Users/shawn.kovalchick/Repositories/contribution/checkstyle-tester/src/main/java/hibernate-orm/documentation/src/main/docbook/devguide/en-US/chapters/query_criteria/extras/select_wrapper.java:13:82: expecting IDENT, found ')'
/Users/shawn.kovalchick/Repositories/contribution/checkstyle-tester/src/main/java/hibernate-orm/documentation/src/main/docbook/devguide/en-US/extras/cache_providers_mapping.java:4:23: expecting RCURLY, found '...'
/Users/shawn.kovalchick/Repositories/contribution/checkstyle-tester/src/main/java/hibernate-orm/documentation/src/main/docbook/devguide/en-US/extras/timestamp_version.java:3:1: expecting RCURLY, found '...'
/Users/shawn.kovalchick/Repositories/contribution/checkstyle-tester/src/main/java/hibernate-orm/documentation/src/main/docbook/devguide/en-US/extras/timestamp_version.java:5:12: unexpected token: Date
/Users/shawn.kovalchick/Repositories/contribution/checkstyle-tester/src/main/java/hibernate-orm/documentation/src/main/docbook/devguide/en-US/extras/version_annotation.java:3:1: expecting RCURLY, found '...'
/Users/shawn.kovalchick/Repositories/contribution/checkstyle-tester/src/main/java/hibernate-orm/documentation/src/main/docbook/devguide/en-US/extras/version_annotation.java:6:12: unexpected token: Integer
/Users/shawn.kovalchick/Repositories/contribution/checkstyle-tester/src/main/java/lombok-ast/test/resources/parseErrors/Bad1.java:21:2: unexpected token: public
/Users/shawn.kovalchick/Repositories/contribution/checkstyle-tester/src/main/java/openjdk8/test/tools/pack200/typeannos/TypeUseTarget.java:39:74: unexpected token: ...
/Users/shawn.kovalchick/Repositories/contribution/checkstyle-tester/src/main/java/openjdk8/test/tools/pack200/typeannos/TypeUseTarget.java:40:8: unexpected token: Object
/Users/shawn.kovalchick/Repositories/contribution/checkstyle-tester/src/main/java/openjdk8/test/tools/pack200/typeannos/TypeUseTarget.java:40:26: unexpected token: String
/Users/shawn.kovalchick/Repositories/contribution/checkstyle-tester/src/main/java/openjdk8/test/tools/pack200/typeannos/TypeUseTarget.java:40:36: unexpected token: [
/Users/shawn.kovalchick/Repositories/contribution/checkstyle-tester/src/main/java/openjdk9/test/javax/xml/bind/jxc/8073519/InputWithError.java:28:10: unexpected token: compile
/Users/shawn.kovalchick/Repositories/contribution/checkstyle-tester/src/main/java/openjdk9/test/tools/pack200/typeannos/TypeUseTarget.java:39:74: unexpected token: ...
/Users/shawn.kovalchick/Repositories/contribution/checkstyle-tester/src/main/java/openjdk9/test/tools/pack200/typeannos/TypeUseTarget.java:40:8: unexpected token: Object
/Users/shawn.kovalchick/Repositories/contribution/checkstyle-tester/src/main/java/openjdk9/test/tools/pack200/typeannos/TypeUseTarget.java:40:26: unexpected token: String
/Users/shawn.kovalchick/Repositories/contribution/checkstyle-tester/src/main/java/openjdk9/test/tools/pack200/typeannos/TypeUseTarget.java:40:36: unexpected token: [
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 05:49 min
[INFO] Finished at: 2017-01-21T16:48:40-05:00
[INFO] Final Memory: 42M/2083M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-site-plugin:3.3:site (default-site) on project sample: Error during page generation: Error rendering Maven report: Failed during checkstyle execution: There are 83 errors reported by Checkstyle 7.5-SNAPSHOT with my_check.xml ruleset. -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-site-plugin:3.3:site (default-site) on project sample: Error during page generation
```

**Comment by bamapookie** [comment]
I needed to clean out the src/main/java directory.  It was only using the exclude from the first entry, but all the files from the previous checks were still there.

Rerunning now.

**Comment by bamapookie** [comment]
One more minor issue: https://github.com/checkstyle/contribution/issues/158

Fixed locally and running again.

**Comment by bamapookie** [comment]
Here is the latest: https://bamapookie.github.io/checkstyle_issue_3733/diff/

Now I can work on the changes @rnveach requested.

**Comment by romani** [comment]
OK, we are waiting for your next update and report

**Comment by romani** [comment]
@bamapookie , ping .

**Comment by bamapookie** [comment]
Sorry.  I will get to work on it this weekend.  Been crazy busy at work
this week. :)

On Sat, Jan 28, 2017, 8:07 AM Roman Ivanov <notifications@github.com> wrote:

@bamapookie <https://github.com/bamapookie> , ping .

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/checkstyle/checkstyle/pull/3737#issuecomment-275847228>,
or mute the thread

**Comment by romani** [comment]
@bamapookie , please let us know if you have problems with reports, tools or ....

**Comment by bamapookie** [comment]
@romani or a babysitter?  :)  I will have a commit today, but I may have a few questions.  I intend to rewrite the check to look at each top-level annotation independently, and for each line, if it is the first thing on the line, then validate the start point.  Does that sound OK?  Does any part of the check as it exists now handle nested annotations, or should I not worry about that?

**Comment by rnveach** [comment]
@bamapookie babysitter?

>  I may have a few questions

Questions are ok, we don't expect each one of us to know everything.

> for each line, if it is the first thing on the line, then validate the start point.

That is how check should work. We only validate what is starting the line. Everything else should not print violations, but might be used to calculate next line's indentation. We can't assume a specific token always starts a line, otherwise this will lead to gaps in validation and cause errors to go missed.

> Does any part of the check as it exists now handle nested annotations, or should I not worry about that?

Looking at an example of nested annotations online, indentation currently requests it be indented correctly if it is not. So yes, we handle nested annotations and we shouldn't break it, unless we are handling a specific case incorrectly somehow. If there was such a case, we should probably make an issue on it.
Example:
````
@ComplexAnnotation(
    @SimpleAnnotation(a="...", b=3)
)
````

**Comment by bamapookie** [comment]
@rnveach j/k I have two young kids which usually take up my time, hence the lack of progress. :)

**Comment by bamapookie** [comment]
Getting these errors with the new indentation check:
```
[checkstyle] [ERROR] /Users/shawn.kovalchick/Repositories/checkstyle/src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheckTest.java:375: 'public' have incorrect indentation level 4, expected level should be 8. [Indentation]
[checkstyle] [ERROR] /Users/shawn.kovalchick/Repositories/checkstyle/src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportControlLoaderTest.java:76: 'public' have incorrect indentation level 4, expected level should be 8. [Indentation]
[checkstyle] [ERROR] /Users/shawn.kovalchick/Repositories/checkstyle/src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportControlLoaderTest.java:101: 'public' have incorrect indentation level 4, expected level should be 8. [Indentation]
```

These are for comments appearing between the annotation and the method.  Should that be allowed?

**Comment by rnveach** [comment]
>  Should that be allowed?
> CustomImportOrderCheckTest.java:375: 'public' have incorrect indentation level 4, expected level should be 8.

If user can write code like it, we have to validate correctly. Check should not print violations on comments but on code only.
This code is properly formatted. Method starts at indentation 4 and annotation is at same indentation.
We don't really expect any violations on checkstyle's own code.

**Comment by bamapookie** [comment]
So should the indentation check ignore those lines?  Are they validated
elsewhere?  Or should those be handled in the checkIndentation method?

**Comment by rnveach** [comment]
If the line doesn't have any java code on it, it should be ignored. If the line doesn't start with java code, but starts with a comment, it shouldn't print any violations.
Comment indentation is handled by [CommentsIndentationCheck](http://checkstyle.sourceforge.net/config_misc.html#CommentsIndentation).

**Comment by romani** [comment]
> We don't really expect any violations on checkstyle's own code.

indentation violations could appear in our own code, but you need to prove that such violations are valid.

**Comment by bamapookie** [comment]
The error isn't in the comments handling code, but I did add a check for this situation there.  Something in my code is not handling the class declaration line properly when there are comments between the annotations and the class declaration.  I am working on that.  Is there a way to add comments in the annotation test input files?  The files expect comments with indentation check parameters, and there doesn't appear to be a way to allow regular comments in them.

**Comment by bamapookie** [comment]
Running regression test now.  Will push and post soon.

**Comment by bamapookie** [comment]
Trying to run the checkstyle-tester tool.  Getting the following error:
```[ERROR] Failed to execute goal org.apache.maven.plugins:maven-site-plugin:3.3:site (default-site) on project sample: Error during page generation: Error rendering Maven report: Failed during checkstyle execution: There are 3 errors reported by Checkstyle 7.6-SNAPSHOT with my_check.xml ruleset. -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-site-plugin:3.3:site (default-site) on project sample: Error during page generation
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:212)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoExecutionException: Error during page generation
	at org.apache.maven.plugins.site.SiteMojo.execute(SiteMojo.java:143)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
	... 20 more
Caused by: org.apache.maven.doxia.siterenderer.RendererException: Error rendering Maven report: Failed during checkstyle execution
	at org.apache.maven.plugins.site.ReportDocumentRenderer.renderDocument(ReportDocumentRenderer.java:233)
	at org.apache.maven.doxia.siterenderer.DefaultSiteRenderer.renderModule(DefaultSiteRenderer.java:319)
	at org.apache.maven.doxia.siterenderer.DefaultSiteRenderer.render(DefaultSiteRenderer.java:135)
	at org.apache.maven.plugins.site.SiteMojo.renderLocale(SiteMojo.java:175)
	at org.apache.maven.plugins.site.SiteMojo.execute(SiteMojo.java:138)
	... 22 more
Caused by: org.apache.maven.reporting.MavenReportException: Failed during checkstyle execution
	at org.apache.maven.plugin.checkstyle.AbstractCheckstyleReport.executeReport(AbstractCheckstyleReport.java:492)
	at org.apache.maven.plugin.checkstyle.CheckstyleReport.executeReport(CheckstyleReport.java:154)
	at org.apache.maven.reporting.AbstractMavenReport.generate(AbstractMavenReport.java:255)
	at org.apache.maven.plugins.site.ReportDocumentRenderer.renderDocument(ReportDocumentRenderer.java:219)
	... 26 more
Caused by: org.apache.maven.plugin.checkstyle.exec.CheckstyleExecutorException: There are 3 errors reported by Checkstyle 7.6-SNAPSHOT with my_check.xml ruleset.
	at org.apache.maven.plugin.checkstyle.exec.DefaultCheckstyleExecutor.executeCheckstyle(DefaultCheckstyleExecutor.java:313)
	at org.apache.maven.plugin.checkstyle.AbstractCheckstyleReport.executeReport(AbstractCheckstyleReport.java:473)
	... 29 more```
I have pulled the latest version as well as rebased my branch against the upstream changes.  Still getting the above error when running launch.groovy.

**Comment by romani** [comment]
Please provide link to your fork of this tool and your changes.
Please provide command line that you run.

**Comment by bamapookie** [comment]
Fork: https://github.com/bamapookie/checkstyle-contribution
Only changes are to launch.sh, my_check.xml, and projects-to-test-on.properties.

Failing command:
```groovy launch.groovy -l projects-to-test-on.properties -c my_check.xml```
Successful command:
```groovy diff.groovy -r /Users/shawn.kovalchick/Repositories/checkstyle -b master -p issue-3733 -c my_check.xml -l projects-to-test-on.properties```

Results of the diff.groovy command are here: https://bamapookie.github.io/checkstyle-diff-issue-3733/

It looks like I still have an NPE somewhere.

**Comment by bamapookie** [comment]
Just pushed the latest changes.  I got rid of the NPE.  Regression test has been updated as well.

https://bamapookie.github.io/checkstyle-diff-issue-3733/

**Comment by rnveach** [comment]
@bamapookie 
> Failing command:
> groovy launch.groovy
> Successful command:
> groovy diff.groovy

This was most likely why you got an exception. `launch.groovy` is no-exception testing. It will stop regression if there are any exceptions, like NPE. `diff.groovy` is difference reporting. It will work through exceptions and print them as violations in the report.
If `launch.groovy` fails, then your code might have a bug in it.

**Comment by bamapookie** [comment]
@rnveach I did have a bug, but it is fixed.  Should I rerun the launch.groovy file, or is the diff sufficient?

**Comment by rnveach** [comment]
Diff is enough, I was just explaining what happened.

**Comment by bamapookie** [comment]
I have some concerns with some of the results (specifically the differences in the hibernate ORM repo beginning with '@').  I am rerunning the diff on a subset of the repos and adding a tab size of 4 to the my_check.xml.  I will post the results alongside the current set.

**Comment by romani** [comment]
@bamapookie ,report is good, thanks a lot !!!

code is ok to merge.

@rnveach , please do final review as you approved code before final fix and reports.

**Comment by romani** [comment]
Good catch! I missed this.

@bamapookie , please squash all commits on one, there is not reason to have them separate.

**Comment by bamapookie** [comment]
I broke something trying to rebase.  Working on it. :(

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LineWrappingHandler.java [hunk 1]
```diff
@@ -187,33 +218,56 @@ private void checkAnnotationIndentation(DetailAST atNode,
...
+            switch (checkNode.getType()) {
+                case TokenTypes.RCURLY:
+                case TokenTypes.RBRACK:
+                    while (checkNode.getNextSibling() == null) {
+                        checkNode = checkNode.getParent();
+                    }
+                    checkNode = checkNode.getNextSibling();
+                    break;
+                default:
+                    return false;
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LineWrappingHandler.java]: single return point from method is required.

Comment (bamapookie) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LineWrappingHandler.java]: Added endOfScope boolean and returned that.

Comment (bamapookie) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LineWrappingHandler.java]: Should single return point be a checkstyle check? :)

Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LineWrappingHandler.java]: yes, it is, we stopped at count=2 just to enforce some level right now.
One day we will refactor all to be count=1  but it is non simple challenge see #3496 .

**Comment by romani** [comment]
wercker execution of checkstyle is hanged on "Postgres jdbc driver" project
I relaunched it , if repeated again - it should be investigated.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LineWrappingHandler.java (hunk 1 above)
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LineWrappingHandler.java]: in previous version there was return on "default" , not it is continue to do "while", refactoring is not correct.

**Comment by bamapookie** [comment]
I added a label on which to break to handle the last issue.

**Comment by rnveach** [comment]
@bamapookie CI is failing. We don't allow labels.
@romani We should add that to checkstyle as a forbidden item.

**Comment by romani** [comment]
config is updated.
@bamapookie , please rebase on our latest master.

**Comment by bamapookie** [comment]
Rebased and removed the label.  Added the boolean to the while loop to break there.

**Comment by romani** [comment]
I feel suspicious about this PR, after that small mess with rebase. 
@bamapookie Please regenerate diff report one more time, to make sure we are merging good code.

@rnveach , please confirm again that you are OK to merge.

**Comment by bamapookie** [comment]
I reran the last check with the tabsize set to 4.  When I went to commit, there was no difference to commit.  I am rerunning the longer, earlier one with no tab size set.

**Comment by bamapookie** [comment]
Updated https://bamapookie.github.io/checkstyle-diff-issue-3733.  The only difference that I see is that I didn't include all the same repos.  The other diff wit tabsize 4 was unchanged.

**Comment by bamapookie** [comment]
I added in the missing three repos.  There are a lot of differences there in the findbugs repo.  @romani @rnveach could you review before approving?  I am going to update the one with tabsize 4 to include findbugs too.

**Comment by bamapookie** [comment]
Tab size 4 diff is uploaded.  It looks like findbugs has some weird formatting rules that are not handled well.

**Comment by romani** [comment]
@bamapookie , reports is fine as for me.
@rnveach , please review and merge if all is fine.

**Pull request merged**